R/render.r

```diff
@@ -422,7 +422,121 @@ data_deps <- function(pkg, depth) {
       )
     )
   )
-  deps <- bslib::bs_theme_dependencies(bs_theme)
+
+  # map either secondary OR fg, bg to component-active-bg
+  # and also dropdown-link-active-bg
+  # unless a value was set by the user
+  if (!is.null(pkg$meta[["template"]]$bslib$secondary)) {
+    if (is.null(pkg$meta[["template"]]$bslib$`component-active-bg`)) {
+      bs_theme <- bslib::bs_add_variables(
+        bs_theme,
+        "component-active-bg" = as.character(
+          bslib::bs_get_variables(bs_theme, "secondary")
+        )
+      )
+    }
+  } else {
+    if (is.null(pkg$meta[["template"]]$bslib$`component-active-bg`)) {
+      bs_theme <- bslib::bs_add_variables(
+        bs_theme,
+        "component-active-bg" = "mix($body-color, $body-bg, 5%)",
+        .where = "declarations"
```

Comment:
One of the downsides of the `.where = "declarations"` approach is that this new value for `$component-active-bg` won't "cascade" to it's dependents (e.g., `$dropdown-link-active-bg`). In this case, it might be cleaner to do something like

```r
cols <- bs_get_variables(theme, c("bg", "fg"))
theme <- bs_theme(theme, "component-active-bg" = sprintf("mix(%s, %s, 5%)", cols$fg, cols$bg))
```

Reply:
Yes, I've been wondering a lot about the `.where = "declarations"`.

Do you have a working example for your pseudo-code?

``` r
library("bslib")
#> 
#> Attachement du package : 'bslib'
#> L'objet suivant est masqué depuis 'package:utils':
#> 
#>     page
theme <- bs_theme(version = 4)
cols <- bs_get_variables(theme, c("bg", "fg"))
theme <- bs_theme(theme, "component-active-bg" = sprintf("mix(%s, %s, 5%)", cols$fg, cols$bg))
#> Error in validate_layer_param(defaults, "defaults"): `sass_layer(defaults)` can not contain another `sass_bundle()` object.
#> Found:
#> /* Sass Bundle: _root, _reboot, _type, _images, _code, _grid, _tables, _forms, _buttons, _transitions, _dropdown, _button-group, _input-group, _custom-forms, _nav, _navbar, _card, _breadcrumb, _pagination, _badge, _jumbotron, _alert, _progress, _media, _list-group, _close, _toasts, _modal, _tooltip, _popover, _carousel, _spinners, _utilities, _print, bs3compat, nav_spacer */
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_deprecated.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_functions.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/sass-utils/color-contrast.scss";
#> 
#> 
#> 
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/bs3compat/_defaults.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_variables.scss";
#> $bootstrap-version: 4;
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_mixins.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/bs3compat/_declarations.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_root.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_reboot.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_type.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_images.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_code.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_grid.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_tables.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_forms.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_buttons.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_transitions.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_dropdown.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_button-group.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_input-group.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_custom-forms.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_nav.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_navbar.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_card.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_breadcrumb.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_pagination.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_badge.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_jumbotron.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_alert.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_progress.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_media.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_list-group.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_close.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_toasts.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_modal.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_tooltip.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_popover.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_carousel.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_spinners.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_utilities.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs4/scss/_print.scss";
#> .table th[align=left] { text-align: left; }
#> .table th[align=right] { text-align: right; }
#> .table th[align=center] { text-align: center; }
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/bs3compat/_rules.scss";
#> @import "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/nav-spacer/nav-spacer.scss";
#> /* *** */
#> 
#> Other Sass Bundle information:
#> List of 2
#>  $ html_deps       :List of 1
#>   ..$ :List of 10
#>   .. ..$ name      : chr "bs3compat"
#>   .. ..$ version   : chr "0.2.5.9002"
#>   .. ..$ src       :List of 1
#>   .. .. ..$ file: chr "bs3compat/js"
#>   .. ..$ meta      : NULL
#>   .. ..$ script    : chr [1:3] "transition.js" "tabs.js" "bs3compat.js"
#>   .. ..$ stylesheet: NULL
#>   .. ..$ head      : NULL
#>   .. ..$ attachment: NULL
#>   .. ..$ package   : chr "bslib"
#>   .. ..$ all_files : logi TRUE
#>   .. ..- attr(*, "class")= chr "html_dependency"
#>  $ file_attachments: Named chr "/home/maelle/R/x86_64-pc-linux-gnu-library/4.1/bslib/lib/bs3/assets/fonts"
#>   ..- attr(*, "names")= chr "fonts"
```

<sup>Created on 2021-07-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Reply:
Change that last line to `bs_add_variables(theme, "component-active-bg" = paste0("mix(", cols[["fg"]], ",", cols[["bg"]], ",", "5%)"))`

Reply:
got it, thank you!